feat(router): update intro, installation, and add linking doc (#23153)
expo/expo: docs/pages/routing/layouts.mdx

---
title: Layout Routes
description: Learn how to define shared UI elements like tab bars and headers.
hideFromSearch: true
---

import { FileTree } from '~/ui/components/FileTree';
import { GithubIcon, BookOpen02Icon, Rocket02Icon } from '@expo/styleguide-icons';
import { BoxLink } from '~/ui/components/BoxLink';
import { CODE } from '~/ui/components/Text';

> By default, routes fill the entire screen. Moving between them is a full-page transition with no animation. In native apps, users expect shared elements like headers and tab bars to persist between pages. These can be created using **layout routes**.

Reviewer: ```suggestion
By default, routes fill the entire screen. Moving between them is a full-page transition with no animation. In native apps, users expect shared elements like headers and tab bars to persist between pages. These are created using **layout routes**.
```

## Create a layout route

To create a layout route for a directory, create a file named **\_layout.js** in the directory, and export a React component as `default`.

```js app/home/_layout.js

> import { Slot } from 'expo-router';

Reviewer: ```suggestion
import { Slot } from 'expo-router';

```

export default function HomeLayout() {
  return <Slot />;
}
```

From the above example, **Slot** will render the current child route, think of this like the `children` prop in React. This component can be wrapped with other components to create a layout.

```js app/home/_layout.js

> import { Slot } from 'expo-router';

Reviewer: ```suggestion
import { Slot } from 'expo-router';

```

export default function HomeLayout() {
  return (
    <>
      <Header />
      <Slot />
      <Footer />
    </>
  );
}
```

Expo Router supports adding a single layout route for a given directory. If you want to use multiple layout routes, add multiple directories:

<FileTree files={['app/_layout.js', 'app/home/_layout.js', 'app/home/index.js']} />

```js app/_layout.js
import { Tabs } from 'expo-router';

export default function Layout() {
  return <Tabs />;
}
```

```js app/home/_layout.js
import { Stack } from 'expo-router';

export default function Layout() {
  return <Stack />;
}
```

If you want multiple layout routes without modifying the URL, you can use [groups](#groups).

> ## Groups

Reviewer: This section overlaps in terms of duplicated info from this section: https://github.com/expo/expo/pull/23153/files#diff-ca58a0cc69465f31947e52981ff3361c0f38e0d959beb4ed78ab00ea886d28b6R11

Maybe, we can have that info in one place and then link to the info from the other place.

You can prevent a segment from showing in the URL by using the group syntax `()`.

- **app/root/home.js** matches `/root/home`
- **app/(root)/home.js** matches `/home`

This is useful for adding layouts without adding additional segments to the URL. You can add as many groups as you want.

Groups are also good for organizing sections of the app. In the following example, we have **app/(app)** which is where the main app lives, and **app/(aux)** which is where auxiliary pages live. This is useful for adding pages which you want to link to externally, but don't need to be part of the main app.

<FileTree
  files={[
    'app/(app)/index.js',
    'app/(app)/user.js',
    'app/(aux)/terms-of-service.js',
    'app/(aux)/privacy-policy.js',
  ]}
/>

## Native layouts

> One of the best advantages to React Native is being able to use native UI components. Expo Router provides a few drop-in native layouts that you can use to easily achieve familiar native behavior. You can change between truly-native layouts on certain platforms and custom layouts on others by following the [Platform-specific modules guide](router/advance/platform-specific-modules).

Reviewer: ```suggestion
One of the best advantages to React Native is being able to use native UI components. Expo Router provides a few drop-in native layouts that you can use to easily achieve familiar native behavior. To change between truly-native layouts on certain platforms and custom layouts on others, see [Platform-specific modules](router/advance/platform-specific-modules).
```
